contrib/babelfishpg_tsql/sql/sys_functions.sql

```diff
@@ -3073,6 +3073,130 @@ CREATE OR REPLACE FUNCTION sys.substring(string sys.NCHAR, i INTEGER, j INTEGER)
 RETURNS sys.NVARCHAR
 AS 'babelfishpg_tsql', 'tsql_varchar_substr' LANGUAGE C IMMUTABLE PARALLEL SAFE;
 
+-- wrapper functions for upper --
+-- Function to handle datatypes which are implicitly convertable to VARCHAR
+CREATE OR REPLACE FUNCTION sys.upper(ANYELEMENT)
+RETURNS sys.VARCHAR
+AS $$
+DECLARE
+    varch sys.varchar;
+    typ_oid oid;
+    typnam text;
+BEGIN
+    typ_oid := (SELECT typbasetype FROM pg_type WHERE oid = pg_typeof($1));
+    typnam := (SELECT typname FROM pg_type WHERE oid in (typ_oid));
+    IF pg_typeof($1) IN ('image'::regtype, 'sql_variant'::regtype, 'xml'::regtype, 'geometry'::regtype, 'geography'::regtype) THEN
+        RAISE EXCEPTION 'Argument data type % is invalid for argument 1 of upper function.', pg_typeof($1);
+    ELSIF typnam IN ( 'image'::regtype::name COLLATE "C", 'sql_variant'::regtype::name COLLATE "C", 'xml'::regtype::name COLLATE "C", 'geometry'::regtype::name COLLATE "C", 'geography'::regtype::name COLLATE "C") THEN
+        RAISE EXCEPTION 'Argument data type % is invalid for argument 1 of upper function.', typnam;
+    END IF;
+    IF $1 IS NULL THEN
+        RETURN NULL;
+    END IF;
+    varch := (SELECT CAST ($1 AS sys.varchar));
+    -- Call the underlying function after preprocessing
+    RETURN (SELECT pg_catalog.upper(varch));
```

Comment:
We don't need to explicit cast here, function return type will implicitly cast it.